### PR TITLE
Working towards display plugins

### PR DESCRIPTION
**Pull request by jherico**
This pulls in some of the dependencies I'm using for the new display plugins code. Specifically, I'm using oglplus in order to simplify rendering of content to the display plugin rendering context.  

You might ask why I'm not using the existing GPU code.  Well, that's because the existing GPU code isn't locked to an OpenGL backend, and may change depending on the platform and the future availability of Vulkan.  more importantly, the GPU backend isn't designed to work with multiple rendering contexts.  

The display plugins take a texture ID for the scene and another for the UI overlay and are responsible for rendering that.  Rather than refactor the GPU code to be cross-context enabled, it's simpler for the display plugins to be responsible for getting those textures to their surfaces with a minimum amount of code and a minimum number of dependencies on our other libraries, hence oglplus.  Boost config is used in order to allow the use of oglplus as a header only library.  Without it, you'd need a cmake initialization step to create the per-platform (mostly per-compiler actually) header that oglplus would use to toggle on and off preprocessor defines.  

Additionally there are some minor cleanups, like consolidating two classes that do the same thing:  OffscreenGlContext and OffscreenGlCanvas (my bad), and updating the version of OpenVR.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 09085f1cd93318d402871b6d2d435723a47f98f9 available at: https://gist.github.com/a91fa7edeb0b7c54b2e5<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4969.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4969.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: e5bc0fad1ec93e3628134547e541c01448a7ca66 available at: https://gist.github.com/a91fa7edeb0b7c54b2e5<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4969.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4969.dmg

File: libraries/octree/src/ViewFrustum.h [hunk 1]
```diff
@@ -54,7 +52,8 @@ class ViewFrustum {
     void getFocalLength(float focalLength) { _focalLength = focalLength; }
 
     // getters for lens attributes
-    const glm::mat4 getProjection() const { return _projection; };
+    const glm::mat4& getProjection() const { return _projection; };
```
Comment (ZappoMan) [inline libraries/octree/src/ViewFrustum.h]: trailing ; not needed

File: libraries/octree/src/ViewFrustum.h [hunk 2]
```diff
@@ -54,7 +52,8 @@ class ViewFrustum {
...
+    const glm::mat4& getView() const { return _view; };
```
Comment (ZappoMan) [inline libraries/octree/src/ViewFrustum.h]: trailing ; not needed

File: libraries/render-utils/src/GlWindow.h [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+class GlWindow : public QWindow {
+public:
+    GlWindow(QOpenGLContext * shareContext = nullptr);
+    GlWindow(const QSurfaceFormat& format, QOpenGLContext * shareContext = nullptr);
+    virtual ~GlWindow();
+    void makeCurrent();
+    void doneCurrent();
+    void swapBuffers();
+private:
+    QOpenGLContext * _context{ nullptr };
```
Comment (ZappoMan) [inline libraries/render-utils/src/GlWindow.h]: type* squish

File: libraries/render-utils/src/GlWindow.h [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+#ifdef DEBUG
+    QOpenGLDebugLogger * _logger{ nullptr };
```
Comment (ZappoMan) [inline libraries/render-utils/src/GlWindow.h]: type* squish

File: libraries/render-utils/src/GlWindow.h [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+#include <QWindow>
+#include <QSurfaceFormat>
+
+class QOpenGLContext;
+class QOpenGLDebugLogger;
+
+class GlWindow : public QWindow {
+public:
+    GlWindow(QOpenGLContext * shareContext = nullptr);
```
Comment (ZappoMan) [inline libraries/render-utils/src/GlWindow.h]: type* squish

File: libraries/render-utils/src/GlWindow.h [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    GlWindow(const QSurfaceFormat& format, QOpenGLContext * shareContext = nullptr);
```
Comment (ZappoMan) [inline libraries/render-utils/src/GlWindow.h]: type* squish

File: libraries/render-utils/src/OffscreenGlCanvas.h [hunk 7]
```diff
@@ -15,16 +15,24 @@
...
     void doneCurrent();
+    QOpenGLContext* getContext() {
+        return &_context;
+    }
 
 protected:
     QOpenGLContext _context;
     QOffscreenSurface _offscreenSurface;
+#ifdef DEBUG
+    QOpenGLDebugLogger * _logger{ nullptr };
```
Comment (ZappoMan) [inline libraries/render-utils/src/OffscreenGlCanvas.h]: type* squish

File: libraries/render-utils/src/GlWindow.cpp [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    format.setStencilBufferSize(8);
+    format.setVersion(4, 1);
+#ifdef DEBUG
+    format.setOption(QSurfaceFormat::DebugContext);
+#endif
+    format.setProfile(QSurfaceFormat::OpenGLContextProfile::CompatibilityProfile);
+    return format;
+}
+
+GlWindow::GlWindow(QOpenGLContext * shareContext) : GlWindow(getDefaultFormat(), shareContext) {
```
Comment (ZappoMan) [inline libraries/render-utils/src/GlWindow.cpp]: type* squish

File: libraries/render-utils/src/GlWindow.cpp [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+GlWindow::GlWindow(const QSurfaceFormat& format, QOpenGLContext * shareContext) {
```
Comment (ZappoMan) [inline libraries/render-utils/src/GlWindow.cpp]: type* squish

File: libraries/render-utils/src/GlWindow.cpp [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+#include "GlWindow.h"
+
+#include <QOpenGLContext>
+#include <QOpenGLDebugLogger>
+
+static QSurfaceFormat getDefaultFormat() {
+    QSurfaceFormat format;
+    // Qt Quick may need a depth and stencil buffer. Always make sure these are available.
+    format.setDepthBufferSize(16);
```
Comment (ZappoMan) [inline libraries/render-utils/src/GlWindow.cpp]: this feels like magic to me... why 16? can we make a constant with a descriptive name?

File: libraries/render-utils/src/GlWindow.cpp [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+    format.setStencilBufferSize(8);
```
Comment (ZappoMan) [inline libraries/render-utils/src/GlWindow.cpp]: magic?

**Comment by hifi-gustavo** [comment]
Build console output for commit: f98305dd555f351db3deb35a9d4752a12e8616d5 available at: https://gist.github.com/a91fa7edeb0b7c54b2e5<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4969.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4969.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 77b45e2b4adbba05f32e5d77092545cd6f033ad0 available at: https://gist.github.com/a91fa7edeb0b7c54b2e5<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4969.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4969.dmg

**Pull request merged**